src/Lykke.Cqrs.MessageCancellation/Services/MessageCancellationService.cs

```diff
@@ -0,0 +1,87 @@
+﻿using Lykke.Cqrs.MessageCancellation.Services.Interfaces;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Lykke.Cqrs.MessageCancellation.Services
+{
+    public class MessageCancellationService : IMessageCancellationService
+    {
+        private ReaderWriterLockSlim _readWriterLockSlim;
```

Comment:
For async code it's better to use SemaphoreSlim instead of ReaderWriterLockSlim - see https://msdn.microsoft.com/en-us/magazine/jj991977.aspx

Reply:
Current code uses only async interfaces, but in fact there is no async operations.
In case when more functionality is required